### PR TITLE
Public rooms

### LINKED ISSUE #19: Add room ID to url so it’s kept on refresh

### LINKED ISSUE #114: Ability to share link to give to people without Nextcloud account
As we discussed, this should be part of 1.0 especially because Skype now even has that ability.

cc @nickvergessen

### LINKED ISSUE #124: spreed-session does not get updated
spreed-session does not get updated here
https://github.com/nextcloud/spreed/blob/public-rooms/lib/Controller/SignallingController.php#L150

after being set here
https://github.com/nextcloud/spreed/blob/public-rooms/lib/Controller/ApiController.php#L492

any idea why this happens? @nickvergessen @LukasReschke

### LINKED ISSUE #125: Undefined offset: 0 at SignallingController.php#181
I got this after the very first start of the app (without joining any room) once every second.

```
{"reqId":"z\/MEXgAtTznlQvkG59bB","remoteAddr":"::1","app":"PHP","message":"Undefined offset: 0 at \/Users\/morrisjobke\/Projects\/nextcloud\/server\/apps\/spreed\/lib\/Controller\/SignallingController.php#181","level":0,"time":"2016-11-22T08:49:20+00:00","method":"GET","url":"\/server\/index.php\/apps\/spreed\/messages?requesttoken=E1umKmZzlYanXYyHkmIONK14gTw1VYNkQJEJ3byZ%2FRg%3D%3AWx%2FrbggH5cOXGuHA2yV9BtsK4nlXENRWD6Jn7NTOtSo%3D","user":"admin","version":"11.0.0.2"}
```

cc @nickvergessen @Ivansss @LukasReschke

### PR DESCRIPTION
**Pull request by nickvergessen**
Room links are `/apps/spreed?roomId=23`
This allows us to use the same link for both, logged in users and guests.
It also allows us to use the `OC.Util.History.pushState` so reloading the page keeps you in the last room.

## ToDo
- [x] Fix guest view

Fix #19 
Fix #114
Fix #124 
Fix #125

### CODE REVIEW

File: lib/Controller/ApiController.php [hunk 1]
```diff
@@ -186,21 +186,44 @@ public function getRooms() {
...
 	 */
 	public function getPeersInRoom($roomId) {
 		try {
-			$room = $this->manager->getRoomForParticipant($roomId, $this->userId);
+			$room = $this->manager->getRoomById($roomId);
 		} catch (RoomNotFoundException $e) {
 			return new JSONResponse([], Http::STATUS_NOT_FOUND);
 		}
 
+		$participants = $room->getParticipants(time() - 10);
```
Comment (LukasReschke) [inline lib/Controller/ApiController.php]: Use `- 30`, -10 is too few as the signalling requests can take some time. I've added some more logic to the JS code tonight that makes -30 a fair choice.

File: lib/Controller/PageController.php [hunk 2]
```diff
@@ -66,4 +113,30 @@ public function index() {
...
+			$room = $this->manager->getRoomById($roomId);
+			if ($room->getType() !== Room::PUBLIC_CALL) {
+				throw new RoomNotFoundException();
+			}
+		} catch (RoomNotFoundException $e) {
+			throw new HintException($this->l10n->t('The room does not exist.'));
+		}
+
+		$params = [
+			'sessionId' => 'lukas-fix-rand',
```
Comment (LukasReschke) [inline lib/Controller/PageController.php]: Can just be a ISecureRandom session ID with 255 characters. :)

**Comment by jancborchardt** [comment]
I still get ` The room does not exist.` even though there’s someone in there. Is this not implemented yet?

**Comment by nickvergessen** [comment]
Well currently you need to manually change the room type, or call the api directly, because the js doesnt do that yet.

TO change it manually in the db change the rooms type to `3`

**Comment by jancborchardt** [comment]
@nickvergessen @Ivansss what’s the state here? Is it ready for testing so we can review and get it merged?

**Comment by nickvergessen** [comment]
Well API-wise it should be okay now, however the JS needs some more work.
Ivan is on it, not sure if he can finish it today...

**Comment by nickvergessen** [comment]
Should add a background job to purge old guests later on.

**Comment by MorrisJobke** [comment]
I will rebase this - conflict is caused by a merged change I did just minutes ago

**Comment by MorrisJobke** [comment]
Rebased to fix the conflict - cc @Ivansss @nickvergessen

File: lib/Controller/PageController.php [hunk 3]
```diff
@@ -128,11 +134,12 @@ protected function guestEnterRoom($roomId) {
...
 		}
 
+		$newSessionId = $this->secureRandom->generate(255);
 		$params = [
-			'sessionId' => 'lukas-fix-rand',
+			'sessionId' => $newSessionId,
 			'roomId' => $roomId,
 		];
-		$response = new TemplateResponse($this->appName, 'index', $params, 'blank');
+		$response = new TemplateResponse($this->appName, 'index-public', $params, 'base');
```
Comment (Ivansss) [inline lib/Controller/PageController.php]: @jancborchardt Could you please take a look to the css of the public page? It only shows the content correctly when you make window width smaller.

**Comment by MorrisJobke** [comment]
@jancborchardt Could you check out the CSS for the guest view ;)

File: lib/Controller/PageController.php (hunk 3 above)
Comment (MorrisJobke) [inline lib/Controller/PageController.php]: @jancborchardt This one

**Comment by jancborchardt** [comment]
Fixed the CSS.

I had some initial issues though:
- [ ] copying the link via copy to clipboard button didn’t work (Chrome) needed to manually copypaste
- [ ] copying should already be triggered when clicking the actual menu entry »Share link« with feedback from tooltip, on the copy icon
- [ ] initially I had lots of white screens as the anonymous guest
- [ ] reloaded some times, and then at some point I had »5 guests« listed in the room even though only one other browser window had been open. We should detect that better

**Comment by MorrisJobke** [comment]
> initially I had lots of white screens as the anonymous guest

I guess this is this firefox issue we had - with chrome I can't reproduce.

**Comment by MorrisJobke** [comment]
> copying the link via copy to clipboard button didn’t work (Chrome) needed to manually copypaste

Works here in chrome and firefox. Anything in the JS error console?

> reloaded some times, and then at some point I had »5 guests« listed in the room even though only one other browser window had been open. We should detect that better

I guess this is this firefox issue we had - with chrome I can't reproduce.

**Comment by Ivansss** [comment]
> Works here in chrome and firefox. Anything in the JS error console?

Are you using macOS @MorrisJobke ? 
Copying the link via copy to clipboard button doesn't work for me on macOS (neither Chrome nor Firefox).
I don't see any JS errors in the console and even the Clipboard returns a success callback.

**Comment by MorrisJobke** [comment]
> Are you using macOS @MorrisJobke ?
> Copying the link via copy to clipboard button doesn't work for me on macOS (neither Chrome nor Firefox).

Yes - I'm using it and it works here fine in Chrome 😕

**Comment by nickvergessen** [comment]
> reloaded some times, and then at some point I had »5 guests« listed in the room even though only one other browser window had been open. We should detect that better

This is something for a new PR, we need to clean up the guests when they didn't ping for some minutes...

**Comment by nickvergessen** [comment]
Nevermind, I added it to this PR.

One thing I noticed is that for guests it shows the "deleted user" avatar (grey circle with "X") when they have their video off.

**Comment by nickvergessen** [comment]
But I'm fine with merging this now

**Comment by jancborchardt** [comment]
Let’s merge it and fix things going forward? What about the tests?

**Comment by Ivansss** [comment]
It is already rebased to current master. I will merge it too.

**Comment by Ivansss** [comment]
Still one missing JS linting error:
```
/drone/src/github.com/nextcloud/spreed/js/views/roomlistview.js
236:24 error 'Clipboard' is not defined no-undef
```

Any idea where to define it?

**Comment by nickvergessen** [comment]
see https://github.com/nextcloud/spreed/commit/ae4f6b146c34da2f8483085eb1612efd076332bf

**Comment by jancborchardt** [comment]
Also fixed the menu layout which broke at some point :+1:

**Comment by jancborchardt** [comment]
Merged so we don’t run into more issues. :) Now we can test with all functionality in place.

You all rock :metal: :)

**Pull request merged**